### PR TITLE
Bulk Deletions - app and api

### PR DESCRIPTION
**Pull request by dogcomplex**
# Overview

Adds endpoints:

- DELETE /api/activities  to delete a list of activities by their id
- POST /api/deleted/activities  to undelete a list of activities by their id

As well as UI for bulk Deletion on the Activities search page.

Deletion is actually soft-deletion, just setting deleted_timestamp to the current time.  Undeletion is more complex - only un-deleting the latest version.  Might want to adapt that into an "Undo" control which effectively bumps the non-deleted record down the list, or just creates a new record with deleted_timestamp NULL.  Either way - too much effort on too inconsequential details.  Same reasoning there's no UI for undeletion yet.

This PR includes the following proposed change(s):

- https://github.com/bcgov/invasivesbc/issues/316

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Tested locally with dev data.  Deletion UI working

## Screenshots

2021-01-26 20_24_46-InvasivesBC - Search

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/src/paths/activities.ts [hunk 1]
```diff
@@ -7,13 +7,15 @@ import { ALL_ROLES } from '../constants/misc';
 import { getDBConnection } from '../database/db';
 import { ActivitySearchCriteria } from '../models/activity';
 import geoJSON_Feature_Schema from '../openapi/geojson-feature-doc.json';
-import { getActivitiesSQL } from '../queries/activity-queries';
+import { getActivitiesSQL, deleteActivitiesSQL, undeleteActivitiesSQL } from '../queries/activity-queries';
```
Comment (sdevalapurkar) [inline api/src/paths/activities.ts]: remove `undeleteActivitiesSQL` as it is not used

File: api/src/paths/deleted/activities.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
+'use strict';
+
+import { RequestHandler } from 'express';
+import { Operation } from 'express-openapi';
+import { SQLStatement } from 'sql-template-strings';
+import { ALL_ROLES } from '../../constants/misc';
+import { getDBConnection } from '../../database/db';
+import { ActivitySearchCriteria } from '../../models/activity';
```
Comment (sdevalapurkar) [inline api/src/paths/deleted/activities.ts]: remove `ActivitySearchCriteria` as it is unused

File: api/src/paths/deleted/activities.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import geoJSON_Feature_Schema from '../../openapi/geojson-feature-doc.json';
```
Comment (sdevalapurkar) [inline api/src/paths/deleted/activities.ts]: same with geoJSON_Feature_Schema

File: api/src/paths/deleted/activities.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+import { getActivitiesSQL, deleteActivitiesSQL, undeleteActivitiesSQL } from '../../queries/activity-queries';
```
Comment (sdevalapurkar) [inline api/src/paths/deleted/activities.ts]: same with getActivitiesSQL and deleteActivitiesSQL

File: app/src/components/search/SearchResultsList.tsx [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import { Box, Button, makeStyles, Theme, Typography } from '@material-ui/core';
 import SearchActivitiesList from 'components/activities-list/SearchActivitiesList';
 import { ISearchActivity } from 'features/home/search/SearchPage';
+import { DatabaseContext } from 'contexts/DatabaseContext';
```
Comment (sdevalapurkar) [inline app/src/components/search/SearchResultsList.tsx]: remove DatabaseContext

File: app/src/components/search/SearchResultsList.tsx [hunk 6]
```diff
@@ -35,14 +41,29 @@ const SearchResultsList: React.FC<ISearchResultsList> = (props) => {
...
+          </Button>
+          <Button
+            className={classes.bulkButton}
+            disabled={editIds.length === 0}
+            variant="contained"
+            color="primary"
+            onClick={() => {
+              invasivesApi.deleteActivities(editIds);
+              setEditIds([]);
+              history.go(0);
```
Comment (sdevalapurkar) [inline app/src/components/search/SearchResultsList.tsx:65]: supernit:

might be able to just say: `history.go()`

**Review by sdevalapurkar — APPROVED** [review]

**Review by sdevalapurkar — CHANGES_REQUESTED** [review]
LGTM, just need to address the code smells and duplication before merge

File: app/src/components/search/SearchResultsList.tsx (hunk 6 above)
Comment (dogcomplex) [reply app/src/components/search/SearchResultsList.tsx:65]: Compiler errors, so apparently not

Comment (sdevalapurkar) [reply app/src/components/search/SearchResultsList.tsx:65]: hmm interesting, maybe does not work with react-router

**Comment by dogcomplex** [comment]
> LGTM, just need to address the code smells and duplication before merge

@sdevalapurkar Got most of them.  Can we squash the rest?  
- Removing the "unnecessary" typing creates new typing errors due to ambiguity and it used in other endpoints.  
- The POST/DELETE operations are used despite its protests - though not sure why those are being flagged while the POST for getActivities is not (any thoughts on that?). 
- Code duplication:  I'm not in the mood to make a generalized SQL query connector/executor just to save a dozen lines of code.  Hard pass for this PR, and the same crime has been committed many times with other endpoints.

**Comment by sdevalapurkar** [comment]
I think these code smells can be fixed: 

- https://sonarcloud.io/project/issues?id=bcgov_invasivesbc&open=AXdFNQDhJdVl9hNwWZaD&pullRequest=473&resolved=false&types=CODE_SMELL
- https://sonarcloud.io/project/issues?id=bcgov_invasivesbc&open=AXdFLdUuBFS7daocb3PX&pullRequest=473&resolved=false&types=CODE_SMELL

**Comment by sdevalapurkar** [comment]
the duplication should be fixed. the quality gate should pass prior to any PR merge -> just a function that peforms this logic:


just needs a param to allow it to determine whether it is deleting or undeleting but the rest of the code is shared

**Review by sdevalapurkar — DISMISSED** [review]

File: app/src/components/search/SearchResultsList.tsx [hunk 7]
```diff
@@ -39,14 +44,29 @@ const SearchResultsList: React.FC<ISearchResultsList> = (props) => {
...
-          {editIds.length ? ' (' + editIds.length + ')' : ''}
-        </Button>
+        <Box className={classes.resultsBar}>
+          <Typography variant="h5">{editIds.length ? editIds.length + ' Selected' : ''}</Typography>
+          <Button
+            className={classes.bulkButton}
+            disabled={editIds.length === 0}
+            variant="contained"
+            color="primary"
+            onClick={() => navigateToBulkEditPage(editIds)}>
```
Comment (sdevalapurkar) [inline app/src/components/search/SearchResultsList.tsx]:
```suggestion
            onClick={() => navigateToEditPage(editIds)}>
```

**Review by sdevalapurkar — APPROVED** [review]

**Pull request merged**